examples/ex42.cpp

```diff
@@ -1034,18 +1119,44 @@ ApplyGradient3D(const int ne, const Array<double> &B_, const Array<double> &G_,
 
    MFEM_FORALL_3D(e, ne, q1d, q1d, q1d,
    {
+      // shared memory 1D Basis functions filled from b and g inputs
+      MFEM_SHARED tensor<double, q1d, d1d> B, G;
+      // shared memory placeholders for temporary contraction results
+      MFEM_SHARED tensor<double, 2, 3, q1d, q1d, q1d> smem;
       // cauchy stress
       MFEM_SHARED tensor<double, q1d, q1d, q1d, dim, dim> invJ_dsigma_detJw;
+      // du/dxi, ddu/dxi
+      MFEM_SHARED tensor<double, q1d, q1d, q1d, dim, dim> dudxi, ddudxi;
 
-      // du/dxi
-      tensor<double, q1d, q1d, q1d, dim, dim> dudxi{};
+      // initializer '{}' not allowed for shared variable
+      MFEM_FOREACH_THREAD(qx, x, q1d)
+      {
+         MFEM_FOREACH_THREAD(qy, y, q1d)
+         {
+            if (MFEM_THREAD_ID(z) == 0 && qy < d1d)
+            {
+               const int dy = qy;
+               B(qx, dy) = b(qx, dy);
+               G(qx, dy) = g(qx, dy);
+            }
+            MFEM_FOREACH_THREAD(qz, z, q1d)
+            {
+               for (int l = 0; l < dim; l++)
+               {
+                  for (int m = 0; m < dim; m++)
+                  {
+                     dudxi(qx, qy, qz, l, m) = 0.0;
+                     ddudxi(qx, qy, qz, l, m) = 0.0;
+                  }
+               }
+            }
+         }
+      }
```

Comment:
Ugh... I really don't like this. `tensor` is supposed to make things easier.

Reply:
This is exactly what I was referring to when I was saying that we want to abstract the parallelization strategy for a `tensor`, this is precisely to avoid this kind of issues. But it should be fairly easy to hide for now.

Reply:
> Ugh... I really don't like this. `tensor` is supposed to make things easier.

It looks like the verbose implementation is a byproduct of trying to use macros to define the CPU and GPU kernels at the same time. e.g. a CPU-only implementation would just be the one-liner:
```cpp
tensor<double, q1d, q1d, q1d, dim, dim> dudxi{};
```
and a GPU-only implementation would be (assuming kernel launched with appropriate block dimensions)
```cpp
__shared__ tensor<double, q1d, q1d, q1d, dim, dim> dudxi;
du_dxi(threadIdx.x, threadIdx.y, threadIdx.z) = tensor<double, dim, dim>{};
```